index.d.ts

```diff
@@ -0,0 +1,147 @@
+export as namespace EventTargetShim;
+
+export interface Event {
+    /**
+     * The type of this event.
+     */
+    readonly type: string;
+
+    /**
+     * The target of this event.
+     */
+    readonly target: EventTarget;
+
+    /**
+     * The target of this event.
+     */
+    readonly currentTarget: EventTarget;
+
+    /**
+     * The composed path of this event.
+     */
+    composedPath(): EventTarget[];
+
+    /**
+     * Constant of NONE.
+     */
+    readonly NONE: number;
+
+    /**
+     * Constant of CAPTURING_PHASE.
+     */
+    readonly CAPTURING_PHASE: number;
+
+    /**
+     * Constant of AT_TARGET.
+     */
+    readonly AT_TARGET: number;
+
+    /**
+     * Indicates which phase of the event flow is currently being evaluated.
+     */
+    readonly eventPhase: number;
+
+    /**
+     * Stop event bubbling.
+     */
+    stopPropagation(): void;
+
+    /**
+     * Stop event bubbling.
+     */
+    stopImmediatePropagation(): void;
+
+    /**
+     * The flag indicating bubbling.
+     */
+    readonly bubbles: boolean;
+
+    /**
+     * The flag indicating whether the event can be canceled.
+     */
+    readonly cancelable: boolean;
+
+    /**
+     * Cancel this event.
+     */
+    preventDefault(): void;
+
+    /**
+     * The flag to indicating whether the event was canceled.
+     */
+    readonly defaultPrevented: boolean;
+
+    /**
+     * The flag to indicating if event is composed.
+     */
+    readonly composed: boolean;
+
+    /**
+     * Indicates whether the event was dispatched by the user agent.
+     */
+    readonly isTrusted: boolean;
+
+    /**
+     * The unix time of this event.
+     */
+    readonly timeStamp: number;
+}
+
+export interface EventListenerOptions {
+    capture?: boolean;
+    passive?: boolean;
+    once?: boolean;
+}
+
+export interface EventTarget {
+    /**
+     * Add a given listener to this event target.
+     * @param eventName The event name to add.
+     * @param listener The listener to add.
+     * @param options The options for this listener.
+     * @returns `true` if the listener was actually added.
+     */
+    addEventListener(
+        eventName: string,
+        listener: (this: this, ev: Event) => any,
+        options?: boolean | EventListenerOptions,
+    ): boolean;
+
+    /**
+     * Remove a given listener from this event target.
+     * @param eventName The event name to remove.
+     * @param listener The listener to remove.
+     * @param options The options for this listener.
+     * @returns `true` if the listener was actually removed.
+     */
+    removeEventListener(
+        eventName: string,
+        listener: (this: this, ev: Event) => any,
+        options?: boolean | EventListenerOptions,
+    ): boolean;
+
+    /**
+     * Dispatch a given event.
+     * @param event The event to dispatch.
+     * @returns `false` if canceled.
+     */
+    dispatchEvent(event: Event | { type: string }): boolean;
```

Comment:
`Event | { type: string }` means that `event` object has to either have all of properties defined in `Event` or to have `type` property which is strictly `string` plus any other property.

With that said, would you like to a `Partial<Event> & { type: string }` type instead of `{ type: string }`? `type` would be required, but every other property would be optional, but would have to match type in `Event` if provided.

Reply:
A difference between `interface` and `type` is about undefined properties. The `type` disallows that objects have undefined properties. On the other hand, `interface` allows it.

https://www.typescriptlang.org/play/index.html#src=const%20a%3A%20Event%20%7C%20%7B%20type%3A%20string%20%7D%20%3D%20%7B%20type%3A%20%22test%22%2C%20foo%3A%201%20%7D%0D%0A

Reply:
Sorry, I was thinking about [this](https://www.typescriptlang.org/play/index.html#src=const%20a%20%3D%20%7B%20type%3A%20%22test%22%2C%20foo%3A%201%20%7D%3B%0D%0Aconst%20b%3A%20Event%20%7C%20%7B%20type%3A%20string%20%7D%20%3D%20a%3B%0D%0A).

Anyway, what about `Partial<Event> & { type: string }` (given that final for will allow any property)?

Reply:
It's fine if the following notation is allowed.

https://www.typescriptlang.org/play/index.html#src=function%20dispatch(event%3A%20Partial%3CEvent%3E%20%26%20%7B%20type%3A%20string%20%7D)%20%7B%0D%0A%20%20%20%20%2F%2F%0D%0A%7D%0D%0A%0D%0Adispatch(%7B%20type%3A%20%22test%22%2C%20foo%3A%201%20%7D)%0D%0A

Reply:
yup
https://www.typescriptlang.org/play/index.html#src=type%20PartialEvent%20%3D%20Partial%3CEvent%3E%20%26%20%7B%0D%0A%20%20%20%20type%3A%20string%3B%0D%0A%20%20%20%20%5Bkey%3A%20string%5D%3A%20any%3B%0D%0A%7D%0D%0A%0D%0Afunction%20dispatch(event%3A%20PartialEvent)%20%7B%0D%0A%20%20%20%20%2F%2F%0D%0A%7D%0D%0A%0D%0Adispatch(%7B%20type%3A%20%22test%22%2C%20foo%3A%201%20%7D)%0D%0Adispatch(%7B%20foo%3A%201%20%7D)%0D%0A

Reply:
Great!